### ISSUE TITLE
Training with lora_finetune_distributed is slower than single_device, profile shows that nccl is causing this problem

### ISSUE THREAD

**Issue by seekerzz** [pytorch/torchtune#2264]
Training with 3090*4 GPU cards, log shows that

> Step 29 | loss:2.2152204513549805 lr:8.699999999999999e-05 tokens_per_second_per_gpu:576.6150512695312 peak_memory_active:11.401659488677979 peak_memory_alloc:11.401659488677979 peak_memory_reserved:21.88671875 
Step 30 | loss:2.2007763385772705 lr:8.999999999999999e-05 tokens_per_second_per_gpu:614.4884033203125 peak_memory_active:11.757219791412354 peak_memory_alloc:11.757219791412354 peak_memory_reserved:21.88671875 
Step 31 | loss:2.175762891769409 lr:9.3e-05 tokens_per_second_per_gpu:583.15869140625 peak_memory_active:11.427355289459229 peak_memory_alloc:11.427355289459229 peak_memory_reserved:21.88671875 
Step 32 | loss:2.211038589477539 lr:9.599999999999999e-05 **tokens_per_second_per_gpu:575.8968505859375** peak_memory_active:11.45946741104126 peak_memory_alloc:11.45946741104126 peak_memory_reserved:21.88671875 

However, single device is much faster, 
> Step 15 | loss:2.327071189880371 lr:4.4999999999999996e-05 tokens_per_second_per_gpu:2558.451904296875 peak_memory_active:16.525158405303955 peak_memory_alloc:16.525158405303955 peak_memory_reserved:19.201171875 
Step 16 | loss:2.317321300506592 lr:4.7999999999999994e-05 tokens_per_second_per_gpu:2479.65087890625 peak_memory_active:15.913136005401611 peak_memory_alloc:15.913136005401611 peak_memory_reserved:19.201171875 
Step 17 | loss:2.2833023071289062 lr:5.1e-05 tokens_per_second_per_gpu:2597.876708984375 peak_memory_active:15.949980735778809 peak_memory_alloc:15.949980735778809 peak_memory_reserved:19.201171875 
Step 18 | loss:2.30037784576416 lr:5.399999999999999e-05 **tokens_per_second_per_gpu:2434.292236328125** peak_memory_active:16.893787384033203 peak_memory_alloc:16.893787384033203 peak_memory_reserved:19.201171875 

My yaml is copied from `tune ls` and using Phi3, just modified some path. 
I also checked the profile output, the nccl speed is very slow for 4 cards,  how to debug the bottleneck? Thank you!

![Image](https://github.com/user-attachments/assets/e3718cd6-758b-49e3-8f7d-1f7b8fc9d4a4)

**Comment by seekerzz** [comment]
I use `tune run  --nnodes 1  --nproc_per_node 4 lora_finetune_distributed --config phi3.yaml` to start training,
phi3.yaml: 
```
# Config for multi-device LoRA finetuning in lora_finetune_distributed.py
# using a Phi3 mini (3.8B) model
#
# This config assumes that you've run the following command before launching
# this run:
#   tune download microsoft/Phi-3-mini-4k-instruct --output-dir /tmp/Phi-3-mini-4k-instruct --hf-token <HF_TOKEN>
#
# To launch on 2 devices, run the following command from root:
#   tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config phi3/mini_lora
#
# You can add specific overrides through the command line. For example
# to override the checkpointer directory while launching training
# you can run:
#   tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config phi3/mini_lora checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
#
# This config works best when the model is being fine-tuned on 2+ GPUs.
# For single device LoRA finetuning please use mini_lora_single_device.yaml
# or mini_qlora_single_device.yaml

output_dir: /home/adm/phi3_finetune # /tmp may be deleted by your system. Change it to your preference.

# Model arguments
model:
  _component_: torchtune.models.phi3.lora_phi3_mini
  lora_attn_modules: ['q_proj', 'v_proj', 'output_proj']
  apply_lora_to_mlp: True
  apply_lora_to_output: False
  lora_rank: 8  # higher increases accuracy and memory
  lora_alpha: 16  # usually alpha=2*rank
  lora_dropout: 0.0

# Tokenizer
tokenizer:
  _component_: torchtune.models.phi3.phi3_mini_tokenizer
  path: /home/adm/Phi-3-mini-4k-instruct/tokenizer.model
  max_seq_len: null

# Checkpointer
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /home/adm/Phi-3-mini-4k-instruct
  checkpoint_files: [
    model-00001-of-00002.safetensors,
    model-00002-of-00002.safetensors
  ]
  recipe_checkpoint: null
  output_dir: ${output_dir}
  model_type: PHI3_MINI
resume_from_checkpoint: False
save_adapter_weights_only: False

# Dataset
dataset:
  _component_: torchtune.datasets.instruct_dataset
  source: csv
  data_files: /home/adm/qwen_finetune/output_cleaned.csv
  split: train
  train_on_input: True
  new_system_prompt: "续写下列小说："
  packed: False
  column_map:
    input: prompt
    output: output
seed: null
shuffle: True

# Fine-tuning arguments
epochs: 1
max_steps_per_epoch: null
batch_size: 2
gradient_accumulation_steps: 8  # Use to increase effective batch size
optimizer:
  _component_: torch.optim.AdamW
  fused: True
  weight_decay: 0.01
  lr: 3e-4
lr_scheduler:
  _component_: torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup
  num_warmup_steps: 100
loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss
compile: False  # torch.compile the model + loss, True increases speed + decreases memory

# Training env
device: cuda

# Memory management
enable_activation_checkpointing: False  # True reduces memory
enable_activation_offloading: False  # True reduces memory
dtype: bf16

# Logging
metric_logger:
  _component_: torchtune.training.metric_logging.DiskLogger
  log_dir: ${output_dir}/logs
log_every_n_steps: 1
log_peak_memory_stats: True


# Profiler (disabled)
profiler:
  _component_: torchtune.training.setup_torch_profiler
  enabled: False

  #Output directory of trace artifacts
  output_dir: ${output_dir}/profiling_outputs

  #`torch.profiler.ProfilerActivity` types to trace
  cpu: True
  cuda: True

  #trace options passed to `torch.profiler.profile`
  profile_memory: False
  with_stack: False
  record_shapes: True
  with_flops: False

  # `torch.profiler.schedule` options:
  # wait_steps -> wait, warmup_steps -> warmup, active_steps -> active, num_cycles -> repeat
  wait_steps: 5
  warmup_steps: 3
  active_steps: 2
  num_cycles: 1

```

**Comment by felipemello1** [comment]
@seekerzz , before we dig more, do you mind doing a quick sanity check? Can you run it with any number of gpus that is not 4?

I remember seeing in the past that, for some weird reason, when ngpus=4, it is much slower. It wouldnt happen with n=2 or n=8, for example. Lets rule that out, and then we can take a look at the config and if there are other issues.

**Comment by joecummings** [comment]
Also, can I ask what version of torchtune you're using?

**Comment by ebsmothers** [comment]
Yeah also worth running `nvidia-smi topo -m` to see the interconnect between your different devices. If e.g. you are doing heavy comms across PCIe host bridges you may see such slowdowns.

**Comment by seekerzz** [comment]
@joecummings @felipemello1 @ebsmothers 
Thank you for your quick response!
I'm using `torch 2.5.1+cu124, torchao 0.7.0, torchtune 0.5.0, NVIDIA-SMI 560.35.05              Driver Version: 560.35.05      CUDA Version: 12.6`
`nvidia-smi topo -m` shows as follow:
```
        GPU0    GPU1    GPU2    GPU3    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      PIX     NODE    NODE    0-23    0               N/A
GPU1    PIX      X      NODE    NODE    0-23    0               N/A
GPU2    NODE    NODE     X      PIX     0-23    0               N/A
GPU3    NODE    NODE    PIX      X      0-23    0               N/A
```
I did following comparing:
1. `tune run lora_finetune_single_device --config phi3.yaml`, 1epoch (76602 steps )needs about **120 hours**
2. `tune run  --nnodes 1  --nproc_per_node 1 lora_finetune_distributed --config phi3.yaml`, 1epoch (76602 steps )needs about **137 hours**
3. `export CUDA_VISIBLE_DEVICES='0,1'; tune run  --nnodes 1  --nproc_per_node 2 lora_finetune_distributed --config phi3.yaml`, 1 epoch (38301 steps) needs about **200 hours**
4. `export CUDA_VISIBLE_DEVICES='0,2'; tune run  --nnodes 1  --nproc_per_node 2 lora_finetune_distributed --config phi3.yaml`, 1 epoch (38301 steps) needs about **120 hours**
5. `export CUDA_VISIBLE_DEVICES='0,1,2,3'; tune run  --nnodes 1  --nproc_per_node 4 lora_finetune_distributed --config phi3.yaml` 1 epoch (19150 steps) needs about **135 hours**

Another change is that for the network reason, I manually download the phi3 checkpoint from huggingface. (not using `tune download`. I think this may not be the reason for training slow down?)

**Comment by felipemello1** [comment]
I am going to run a trace of phi3 using single device and distributed and see how long two training steps are taking.

**Comment by seekerzz** [comment]
Thanks for your experiments! I will try other server that has different GPU settings to debug wether this problem is related about torchtune.

**Comment by chuangzhidan** [comment]
> Thanks for your experiments! I will try other server that has different GPU settings to debug wether this problem is related about torchtune.

what did u find eventually?
